tracer/src/Datadog.Trace/Agent/TraceSamplers/RareSampler.cs

```diff
@@ -0,0 +1,71 @@
+// <copyright file="RareSampler.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+
+namespace Datadog.Trace.Agent.TraceSamplers
+{
+    internal class RareSampler : ITraceSampler
+    {
+        private const string RareKey = "_dd.rare";
+        private readonly HashSet<StatsAggregationKey> _keys = new();
+
+        public bool Sample(ArraySegment<Span> trace)
+        {
+            bool priorityGreaterThan0 = trace.Array[trace.Offset].Context.TraceContext.SamplingPriority is int p && p > 0;
+            return priorityGreaterThan0 switch
+            {
+                true => HandlePriorityTrace(trace),
+                false => HandleTrace(trace)
+            };
+        }
+
+        private bool HandlePriorityTrace(ArraySegment<Span> trace)
+        {
+            for (int i = 0; i < trace.Count; i++)
+            {
+                var span = trace.Array[i + trace.Offset];
+                if (span.IsTopLevel || span.GetMetric(Tags.Measured) == 1.0 || span.GetMetric(Tags.PartialSnapshot) > 0)
+                {
+                    // Do not return immediately, as we might have multiple spans in this chunk that are rare
+                    var key = StatsAggregator.BuildKey(span);
+                    _keys.Add(key);
+                }
+            }
+
+            return false;
+        }
+
+        private bool HandleTrace(ArraySegment<Span> trace)
+        {
+            bool sampled = false;
```

Comment:
nit: I find `sampled` a problematic name, does that mean it will be kept or dropped 😅  It's a me-thing, but I just can't get my head around it 🤦‍♂️

Reply:
I changed this to `rareSpanFound` in the latest commits, and then we use that as the return value